src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/ProjectFileClassifier.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using System;
+
+namespace Microsoft.VisualStudio.ProjectSystem
+{
+    internal sealed class ProjectFileClassifier
+    {
+        private readonly string _windows;
+        private readonly string _programFiles86;
+        private readonly string? _programFiles64;
+        private readonly string? _vsInstallationDirectory;
+
+        private string[] _nuGetPackageFolders = Array.Empty<string>();
+        private string? _nuGetPackageFoldersString;
+
+        /// <summary>
+        /// Gets and sets the <c>MSBuildProjectExtensionsPath</c> property value for this project.
+        /// Project files under this folder are considered non-modifiable.
```

Comment:
Are you referring to the addition of these files, or modification in place?

The assumption behind this classifier is that once a file is present here, it is not modified in place. It sounds like you're saying that modifications here can happen. Could you provide more details on this?

Reply:
Any of those (addition, deletion, modification) are possible; these files are an output of the NuGet "restore" operation just like project.assets.json. I expect that for the purposes of the Fast Up-to-Date Check we would treat them the same as the project.assets.json.

Reply:
I want to understand how a restore could modify an existing file in any of these locations. What's an example?

The assumptions here are:

- A restore would only add to the NuGet cache, never modify
- A restore would not modify anything in Program Files
- A restore that modifies the assets file will trigger a build (I need to verify this)

Reply:
Possible scenarioes:

1. You update a `PackageReference` to a newer version which includes new .props and/or .targets files in the package. The the .props and .targets generated by NuGet to `$(MSBuildProjectExtensionsPath)` will be updated to import the .props and .targets in the package.
2. A change to the `IncludeAssets`, `ExcludeAssets`, or `PrivateAssets` metadata on a `PackageReference` might change the generated .props/.targets.
3. A change to your nuget.config file causes subsequent restore operations to pick a different version of a package. The generated .props and .targets update as above.
4. If project _B_ references project _A_, a change in a `PackageReference` in _A_ may cause the .props and .targets in _B_ to be generated differently (because of the transitive nature of `PackageReferences`).

I'm not sure any of these are actually an issue for the Fast Up-to-Date Check. If the FUTDC already tracks all the files that are inputs to the restore operation, then perhaps we do not need to be tracking all of the outputs of that restore operation.

Reply:
Another scenario: user selectively deletes a generated .props/.targets from `$(MSBuildProjectExtensionsPath)`. What should the FUTDC do then?

Reply:
As discussed in last week's CPS meeting, we need two classifications.

Files in `MSBuildProjectExtensionsPath` (i.e. the `obj` folder, essentially) can change in place over time. They must be included in the FU2D check, but should not be user editable (these are generated files).

We will introduce a test for "can this file change over time, after being brought into existence" which excludes `MSBuildProjectExtensionsPath`.